### PR TITLE
2963 Update pandera validator templates

### PR DESCRIPTION
**Pull request by tfliss**
This is part of a group of PRs that factors a schema transform out of Pandera/PolaRS validation time. https://github.com/linkml/linkml/issues/2963

This PR specifically:

- removes outdated 'on-the-fly' transform classes and code that have previously been replaced.
- Updates the pandera validator use new cleaner dict->list transform classes.
- updates the pandera schema templates with fixes discovered due to improved dataframe loading via Polars in the compliance tests.
- generates and compiles the polars and pandera models as a combined package in the compliance tests (see note below).
- uses the polars schema forms to load and transform the object to validate
- if pandera testing is selected, validates the polars object, otherwise just load the polars object and check that it is unchanged when written back to json.

Note there are five total schemas in the generated package. Two are pandera and polars schemas that represent dict collections inefficiently as polars 'objects'. These can read the serialized json form directly. Two others are pandera and polars and schemas that only use efficient list collections. The last is a loader 'schema' that transforms the inefficient form into the efficient form.

To run the updated compliance tests (with verbose logging):
`uv run pytest -s -vv --log-cli-level=INFO -m panderagen tests/linkml/test_compliance`
`uv run pytest -s -vv --log-cli-level=INFO -m dataframe_polars_schema tests/linkml/test_compliance`

existing pandera generator tests can be run:
`uv run pytest -s -vv --log-cli-level=INFO tests/linkml/test_generators/panderagen`

To see an example of the generated schema package:

```
uv run gen-pandera --package person examples/PersonSchema/personinfo.yaml
ls person
```

### CODE REVIEW

File: packages/linkml/src/linkml/generators/panderagen/panderagen_class_based/custom_checks.jinja2 [hunk 1]
```diff
@@ -5,17 +5,20 @@
...
+        return cls._check_simple_dict(data, {{ field.reference_class }}, '{{ field.inline_id_column_name }}', '{{ field.inline_other_column_name }}', pa_pl.{{ field.reference_class }}, pa_pl.{{ field.reference_class }}Dict)
         {% elif field.inline_form == 'inline_collection_dict' -%}
-        return cls._check_collection_struct(data)
+        return cls._check_collection_struct(data, {{ field.reference_class }}, pa_pl.{{ field.reference_class }}Struct)
+        {% elif field.inline_form == 'inline_dict' %}
+        return cls._check_nested_struct(data, {{ field.reference_class }}, pa_pl.{{ field.reference_class }})
         {% else -%}
-        return cls._check_nested_list_struct(data)
+        ## NO MATCH FOUND FOR INLINE FORM "{{ field.inline_form }}"
+        raise Exception("No match found during template generation for inline_form: {{ field.inline_form }}")
```
[the remark this answers is not in the document]
Comment (tfliss) [reply packages/linkml/src/linkml/generators/panderagen/panderagen_class_based/custom_checks.jinja2]: The exception is in a macro, and jinja2 doesn't really have a raise built in.
I'll add a raise_ method to field and call that at runtime.

**Comment by tfliss** [comment]
@amc-corey-cox Just an FYI I noticed during the copilot review that I can re-enable more of the nesting unit tests. I might do that as a separate PR, but most likely will put it in this one since the tests are relevant to the functionality. Should be ready for review by early next week.

File: packages/linkml/src/linkml/generators/panderagen/panderagen_polars_schema/load_transformer.jinja2 [hunk 2]
```diff
@@ -13,7 +13,7 @@ from linkml.generators.panderagen.transforms import CollectionDictLoader, Simple
                     struct_schema=loaded_schema.{{ field.reference_class }}Dict,
                     id_col="{{ field.inline_id_column_name }}",
                     other_col="{{ field.inline_other_column_name }}",
```
[the remark this answers is not in the document]
Comment (tfliss) [reply packages/linkml/src/linkml/generators/panderagen/panderagen_polars_schema/load_transformer.jinja2:15]: Ok, I went ahead and passed the calculated id dtype for inline simple dicts to the template

File: tests/linkml/test_compliance/dataframe_helper.py [hunk 3]
```diff
@@ -41,55 +38,77 @@ def apply_skip_list(skip_value: str, skip_list: list[str]) -> None:
...
+        dataframe_serialized_form = pl.from_dicts([object_to_validate], schema=pl_schema, strict=strict_schema)
+
+        actual_data = dataframe_serialized_form.to_dicts()[0]
+        same = deep_compare_dicts(object_to_validate, actual_data)
+
+        logger.info(f"Actual: {actual_data}")
+        logger.info(f"Same: {same}")
+
+        if same and not valid:
+            logger.warning("PolaRS schema accepted an invalid object. Note the schema is not a full validator.")
```
[the remark this answers is not in the document]
Comment (tfliss) [reply tests/linkml/test_compliance/dataframe_helper.py:90]: The polars schema is not designed to validate these forms, so there is nothing to mask.

File: packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py [hunk 4]
```diff
@@ -1,186 +1,147 @@
...
 
-        df = ListDictModelTransform.prepare_dataframe(data, column_name, nested_cls)
+        nested_lf = one_column_lf.explode(column_name).unnest(column_name)
 
-        list_transform = ListDictModelTransform(nested_cls.to_schema())
-        df = list_transform.explode_unnest_dataframe(df, column_name, data)
+        if isinstance(data.lazyframe, pl.LazyFrame):
+            nested_cls.validate(nested_lf)
+        else:
+            nested_cls.validate(nested_lf.collect(engine="streaming"))
```
[the remark this answers is not in the document]
Comment (tfliss) [reply packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py:91]: The name is misleading, it can be a dataframe, and if it is the nested columns should also be collected for consistency.

File: packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py [hunk 5]
```diff
@@ -1,186 +1,147 @@
...
-                                polars_schema[column_name] = pl.List
-                else:
-                    polars_schema[column_name] = dtype.type
-
-        return polars_schema
+        nested_lf = lf.select(column_name).unnest(column_name)
+        if isinstance(data.lazyframe, pl.LazyFrame):
+            nested_cls.validate(nested_lf)
+        else:
+            nested_cls.validate(nested_lf.collect(engine="streaming"))
```
[the remark this answers is not in the document]
Comment (tfliss) [reply packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py:145]: The name is misleading, it can be a dataframe

File: packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py [hunk 6]
```diff
@@ -1,186 +1,147 @@
...
         column_name = data.key
-        nested_cls = cls.get_nested_range(column_name)
 
-        df = NestedStructModelTransform.prepare_dataframe(data, column_name, nested_cls)
-        nested_transform = NestedStructModelTransform(nested_cls.to_schema())
-        df = nested_transform.explode_unnest_dataframe(df, column_name)
+        one_column_lf = data.lazyframe.select(column_name)
+
+        # may be expensive, consider making it optional
+        actual = one_column_lf.collect_schema()[column_name]
```
[the remark this answers is not in the document]
Comment (tfliss) [reply packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py:103]: This is the best way I know of to do the check at present.

File: tests/linkml/test_compliance/dataframe_helper.py [hunk 7]
```diff
@@ -41,55 +38,77 @@ def apply_skip_list(skip_value: str, skip_list: list[str]) -> None:
...
-                    logger.warning("Polars schema accepted an invalid object. Note the schema is not a full validator.")
-                else:
-                    logger.info(f"polars round-trip: {object_to_validate} == {actual_data}")
-            else:
-                if valid:
-                    assert same
-
-            # in a future PR this will fall through to pandera when polars_only is False
-            return
+            assert same
```
[the remark this answers is not in the document]
Comment (tfliss) [reply tests/linkml/test_compliance/dataframe_helper.py:93]: Ok, I addressed this by checking the valid flag and re raising in all cases in the exception handler.

**Comment by tfliss** [comment]
@amc-corey-cox Greetings and happy 2026. just checking if you're still out of office, which I heard might be the case. This is the last of the pandera polars PRs aside from the documentation update, should be ready to go. Also glad to find someone else to ask if you're busy. Thanks!  cc. @sagehrke

**Comment by sagehrke** [comment]
Hey @tfliss! Corey is indeed out this week still. If @kevinschaper or @matentzn would like to step in on this PR, that works too!

**Comment by tfliss** [comment]
> Hey @tfliss! Corey is indeed out this week still. If @kevinschaper or @matentzn would like to step in on this PR, that works too!

Thanks @sagehrke. @kevinschaper, @matentzn I'm glad to answer any questions about this. Most of it is deleting old dict->list transforms, and using new cleaner ones instead (that have been previously unit tested).  Then I update the compliance test helper to load through the first-class polars schema (also previously unit tested). I re-enabled additional unit tests around the dict collection forms. Breaking out the one pandera schema into several simpler interdependent schemas uncovered some edge-case issues in the templates, and these are fixed in this PR as well. There's also a link to the updated documentation in the description of https://github.com/linkml/linkml/pull/2967

File: packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py [hunk 8]
```diff
@@ -1,186 +1,147 @@
...
+            polars_schema_dict[id_col],
+            polars_schema_dict[other_col],
+        )
 
-        one_column_df = data.lazyframe.select(pl.col(column_name)).collect()
+        one_column_lf = data.lazyframe.select(tx.load(column_name)).filter(pl.col(column_name).list.len() > 0)
 
-        list_of_structs = [simple_dict_transformer.transform(e) for [e] in one_column_df.iter_rows()]
+        # may be expensive, consider making it optional
+        actual = one_column_lf.collect_schema()[column_name]
```
[the remark this answers is not in the document]
Comment (tfliss) [reply packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py:49]: Schema collection is really required for this particular case.
It is a lazy evaluation branch, but the library can't determine the schema of the transformed data without collecting at some point. It might be future work to find some other approach but not right now.

File: packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py [hunk 9]
```diff
@@ -1,186 +1,147 @@
...
-        column_name = data.key
+        nested_lf = one_column_lf.explode(column_name).unnest(column_name)
 
-        polars_schema = cls.get_nested_range(column_name).to_schema()
-        simple_transform = SimpleDictModelTransform(polars_schema, *cls._simple_dict_fields(column_name))
-        df = simple_transform.explode_unnest_dataframe(df, column_name)
+        if isinstance(data.lazyframe, pl.LazyFrame):
+            nested_cls.validate(nested_lf)
+        else:
+            nested_cls.validate(nested_lf.collect(engine="streaming"))
```
[the remark this answers is not in the document]
Comment (amc-corey-cox) [reply packages/linkml/src/linkml/generators/panderagen/linkml_pandera_validator.py:62]: It appears to me that this issue still exists. I think it happens in a couple of other places, too. I don't think this needs to be blocking.
